feat(vm): allow attaching existing public IP to VM

wadhah101 opened the pull request:
## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update



@claranet/fr-azure-reviewers

wadhah101 committed:
feat(vm): allow attaching existing public IP to VM

jmapro commented on r-network.tf:
```diff
@@ -1,5 +1,17 @@
+data "azurerm_public_ip" "public_ip" {
+  count = var.public_ip_id == null ? 0 : 1
+
+  name                = split("/", var.public_ip_id)[8]
```
Please create a new local `custom_public_ip_adress_id_parsed` and parse it with https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/functions/parse_resource_id
Then use, `local.custom_public_ip_address_id_parsed.(resource_name|resource_group_name)`

wadhah101 replied:
thanks, i was not aware terraform functions were a thing

updated minimum terraform to 1.8
done

jmapro commented on variables-vm.tf:
```diff
@@ -147,6 +147,12 @@ variable "public_ip_enabled" {
   nullable    = false
 }
 
+variable "public_ip_id" {
```
```suggestion
variable "custom_public_ip_address_id" {
```

wadhah101 replied:
done

jmapro commented on r-network.tf:
```diff
@@ -1,5 +1,17 @@
+locals {
+  public_ip_grp = concat(azurerm_public_ip.main[*], data.azurerm_public_ip.public_ip[*])
```
```suggestion
  public_ip = concat(azurerm_public_ip.main[*], data.azurerm_public_ip.public_ip[*])
```

BzSpi replied:
locals must be in the `locals.tf` file.

wadhah101 replied:
done

jmapro commented on outputs.tf:
```diff
@@ -20,22 +20,22 @@ output "hostname" {
 output "resource_public_ip" {
   description = "VM Public IP resource object."
-  value       = one(azurerm_public_ip.main[*])
+  value       = one(local.public_ip_grp[*])
```
```suggestion
  value       = one(local.public_ip[*])
```

wadhah101 replied:
done

jmapro commented on outputs.tf:
```diff
@@ -20,22 +20,22 @@ output "hostname" {
 output "public_ip_address" {
   description = "Public IP address of the Virtual Machine."
-  value       = one(azurerm_public_ip.main[*].ip_address)
+  value       = one(local.public_ip_grp[*].ip_address)
```
```suggestion
  value       = one(local.public_ip[*].ip_address)
```

wadhah101 replied:
done

jmapro commented on outputs.tf:
```diff
@@ -20,22 +20,22 @@ output "hostname" {
 output "public_ip_id" {
   description = "Public IP ID of the Virtual Machine."
-  value       = one(azurerm_public_ip.main[*].id)
+  value       = one(local.public_ip_grp[*].id)
```
```suggestion
  value       = one(local.public_ip[*].id)
```

wadhah101 replied:
done

jmapro commented on outputs.tf:
```diff
@@ -20,22 +20,22 @@ output "hostname" {
 output "public_domain_name_label" {
   description = "Public DNS of the Virtual Machine."
-  value       = one(azurerm_public_ip.main[*].domain_name_label)
+  value       = one(local.public_ip_grp[*].domain_name_label)
```
```suggestion
  value       = one(local.public_ip[*].domain_name_label)
```

wadhah101 replied:
done

jmapro commented on r-network.tf:
```diff
@@ -1,5 +1,17 @@
+
+
+data "azurerm_public_ip" "public_ip" {
+  count = var.public_ip_id == null ? 0 : 1
```
```suggestion
  count = var.custom_public_ip_address_id == null ? 0 : 1
```

wadhah101 replied:
done

BzSpi commented on r-network.tf:
```diff
@@ -1,5 +1,17 @@
+}
+
+
+data "azurerm_public_ip" "public_ip" {
```
Why using a data source if we already have an ID provided ? This will lead to have a data source possibly evaluated at the apply stage instead of plan stage and this may lead to perpetual drift.

wadhah101 replied:
We need a data source to provide the module outputs with values
`output.public_ip_address` 
and `output.public_domain_name_label`

BzSpi commented on variables-vm.tf:
```diff
@@ -147,6 +147,12 @@ variable "public_ip_enabled" {
+  type        = string
+  default     = null
+}
+
```
Use an object so that the `for_each` can be evaluated even if ID is dynamic. See what's done for `availability_set` variable for example.

```suggestion
variable "public_ip {
  description = "Public IP to attach to the Virtual Machine. if not provided, a new Public IP will be created."
  type = object({
    id = string
  })
  default     = null
}
```

wadhah101 replied:
done

wadhah101 committed:
refactor: rename public_ip_id to custom_public_ip_address_id

Update related resources and outputs
refactor: change server port variable to uppercase and support env
docs: update README tables for public IP fields
chore: bump terraform required_version to >= 1.8

jmapro reviewed (approved):
LGTM

wadhah101 committed:
refactor: rename custom_public_ip_address_id to ip_address_id

wadhah101 committed:
refactor: rename custom_public_ip_address_id to custom_public_ip_address

Shr3ps commented on versions.tf:
```diff
@@ -1,3 +1,3 @@
 terraform {
-  required_version = ">= 1.3"
+  required_version = ">= 1.8"
```
Change to rollback, we already have this constrain in the .tofu file for tofu only

wadhah101 replied:
done

Shr3ps commented on locals.tf:
```diff
@@ -1,4 +1,7 @@
   backup_resource_group_name = var.backup_policy != null ? split("/", var.backup_policy.id)[4] : null
   backup_recovery_vault_name = var.backup_policy != null ? split("/", var.backup_policy.id)[8] : null
+
+  custom_public_ip_adress_parsed = provider::azurerm::parse_resource_id(var.custom_public_ip_address.id)
```
```suggestion
  custom_public_ip_adress_parsed = try(provider::azurerm::parse_resource_id(var.custom_public_ip_address.id), null)
```

wadhah101 replied:
done

Shr3ps commented:
Thanks for your contribution!

wadhah101 committed:
refactor: make custom_public_ip_address parsing safe
lower Terraform required version

Shr3ps commented:
If you can open the same PR and changes on our Windows VM module, would be very cool! Thanks.
